### PR TITLE
Make ladder map selection less random

### LINKED ISSUE #43: Make random map selection in matchmaker less random
It's well known that if people's expectation of "randomness" doesn't match what randomness is in reality.

Make "random map" in matchmaker less random by not repeating the previous 2 maps of all participants **if possible** (don't starve if there is no map that nobody has recently played).

### PR DESCRIPTION
**Pull request by 1-alex98**
Fixes #43

### CODE REVIEW

**Comment by 1-alex98** [comment]
SELECT game_stats.mapId FROM game_player_stats JOIN game_stats WHERE playerId=?0 and game_stats.gameType=6 ORDER BY game_stats.startTime ASC LIMIT ?1

it seems to not like the LIMIT part in the querry but why?

**Comment by bukajsytlos** [comment]
it looks like you are trying to use native sql query instead of jpql

File: src/main/java/com/faforever/server/map/Ladder1v1MapRepository.java [hunk 1]
```diff
@@ -2,9 +2,13 @@
...
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
 
+import java.util.List;
+
 @Repository
 public interface Ladder1v1MapRepository extends JpaRepository<MapVersion, Integer> {
-
+  @Query("SELECT game_stats.mapId FROM game_player_stats JOIN game_stats WHERE playerId=?0 and game_stats.gameType=6 ORDER BY game_stats.startTime ASC LIMIT ?1")
+  List<MapVersion> findLastestLapperMapVersions(int playerId, int number);
```
Comment (micheljung) [inline src/main/java/com/faforever/server/map/Ladder1v1MapRepository.java]: Yes, this is a native query. I doubt this works? You have to use JPQL.

And what is "Lapper"? :P
I'd call it `findRecentlyPlayedLadderMapVersions`

File: src/main/java/com/faforever/server/map/MapService.java [hunk 2]
```diff
@@ -38,9 +40,18 @@ public MapService(MapVersionRepository mapVersionRepository,
...
+    List<MapVersion> modifiedLadderMapPool = new ArrayList<>(originalLadderMapPool);
+
+    modifiedLadderMapPool.removeAll(ladder1v1MapRepository.findLastestLapperMapVersions(host.getId(), 2));
+    modifiedLadderMapPool.removeAll(ladder1v1MapRepository.findLastestLapperMapVersions(opponent.getId(), 2));
+
+    if (modifiedLadderMapPool.size() > 1) {
+      return modifiedLadderMapPool.get(random.nextInt(modifiedLadderMapPool.size() - 1));
+    } else {
+      return originalLadderMapPool.get(random.nextInt(originalLadderMapPool.size() - 1));
+    }
```
Comment (micheljung) [inline src/main/java/com/faforever/server/map/MapService.java]: `else` block is redundant.

**Review by micheljung — DISMISSED** [review]

File: src/main/java/com/faforever/server/map/Ladder1v1MapRepository.java (hunk 1 above)
Comment (1-alex98) [reply src/main/java/com/faforever/server/map/Ladder1v1MapRepository.java]: Why do you doubt it?

Comment (1-alex98) [reply src/main/java/com/faforever/server/map/Ladder1v1MapRepository.java]: oh ok JPQL

File: src/main/java/com/faforever/server/map/Ladder1v1MapRepository.java [hunk 3]
```diff
@@ -1,10 +1,14 @@
...
+import org.springframework.data.domain.Pageable;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
 
 @Repository
 public interface Ladder1v1MapRepository extends JpaRepository<MapVersion, Integer> {
-
+  @Query("SELECT game.mapVersion FROM Game game JOIN game.playerStats playerStats WHERE playerStats.player.id=?0 AND game.featuredMod.id=?1 ORDER BY game.startTime DESC")
+  Page<MapVersion> findRecentlyPlayedLadderMapVersions(Pageable pageble, int playerId, int ladderModId);
```
Comment (micheljung) [inline src/main/java/com/faforever/server/map/Ladder1v1MapRepository.java]: Typo in `pageble`

File: src/main/java/com/faforever/server/map/MapService.java [hunk 4]
```diff
@@ -38,9 +50,28 @@ public MapService(MapVersionRepository mapVersionRepository,
...
+
+    int lastMapsNotConsidered = serverProperties.getLadder1v1().getLastMapsNotConsidered();
+    modifiedLadderMapPool.removeAll(getRecentlyPlayedLadderMapVersions(host, lastMapsNotConsidered));
+    modifiedLadderMapPool.removeAll(getRecentlyPlayedLadderMapVersions(opponent, lastMapsNotConsidered));
+
+    if (modifiedLadderMapPool.size() > 1) {
+      return modifiedLadderMapPool.get(random.nextInt(modifiedLadderMapPool.size() - 1));
+    }
+    return originalLadderMapPool.get(random.nextInt(originalLadderMapPool.size() - 1));
+
```
Comment (micheljung) [inline src/main/java/com/faforever/server/map/MapService.java]: Superfluous Newline :P

**Review by micheljung — CHANGES_REQUESTED** [review]

**Review by micheljung — APPROVED** [review]

**Pull request merged**